### PR TITLE
[GEOT-6393] Add ArcGIS TPK file format reader

### PR DESCRIPTION
**Pull request by leif81**
This adds a new "tpk" module. Authored by my colleague @murraymcdonald16

This is the pull request for this JIRA issue:

https://osgeo-org.atlassian.net/projects/GEOT/issues/GEOT-6393

### CODE REVIEW

File: modules/unsupported/pom.xml [hunk 1]
```diff
@@ -1,237 +1,238 @@
...
-		<module>process-feature</module>
-		<module>geojson</module>
-		<module>mbtiles</module>
-		<module>wfs-ng</module>
-		<module>ogr</module>
-		<module>postgis-raster</module>
-	</modules>
-
-</project>
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (aaime) [inline modules/unsupported/pom.xml]: Please don't rewrite the entire file (it's newlines I think)

File: modules/unsupported/tpk/pom.xml [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.geotools</groupId>
+            <artifactId>gt-coverage</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.geotools</groupId>
+            <artifactId>gt-epsg-hsql</artifactId>
+            <version>${project.version}</version>
```
Comment (aaime) [inline modules/unsupported/tpk/pom.xml]: This should be a test dependency, downstream users should be able to use whatever EPSG plugin they prefer.

File: modules/unsupported/tpk/src/main/java/org/geotools/tpk/TPKBundle.java [hunk 3]
```diff
@@ -0,0 +1,173 @@
+package org.geotools.tpk;
```
Comment (aaime) [inline modules/unsupported/tpk/src/main/java/org/geotools/tpk/TPKBundle.java]: Copyright header missing: please add in all files.

Comment (leif81) [reply modules/unsupported/tpk/src/main/java/org/geotools/tpk/TPKBundle.java]: Ok copyright added now.

File: modules/unsupported/tpk/pom.xml (hunk 2 above)
Comment (leif81) [reply modules/unsupported/tpk/pom.xml]: Ok scope test added now.

File: modules/unsupported/pom.xml (hunk 1 above)
Comment (leif81) [reply modules/unsupported/pom.xml]: Ok this is fixed.

**Comment by leif81** [comment]
@aaime We're looking into resizing the samples to be <100KB and we'll look into your question about the CLA.

**Comment by aaime** [comment]
@leif81 thank you. Let me know when you have signed and sent the CLA to OSGeo, I don't get notified myself (just checked the database of CLAs, see none matching the names of people involved in this PR, so guessing it's still pending)

**Comment by aaime** [comment]
Btw, there is a legit failure in the QA build:

```
[WARNING] COMPILATION WARNING : 
[INFO] -------------------------------------------------------------
[WARNING] /home/travis/build/geotools/geotools/modules/unsupported/tpk/src/main/java/org/geotools/tpk/TPKZoomLevelV2.java:[82,26] Double(double) in java.lang.Double has been deprecated
[WARNING] /home/travis/build/geotools/geotools/modules/unsupported/tpk/src/main/java/org/geotools/tpk/TPKZoomLevelV1.java:[87,26] Double(double) in java.lang.Double has been deprecated
[INFO] 2 warnings 
[INFO] -------------------------------------------------------------
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/travis/build/geotools/geotools/modules/unsupported/tpk/src/main/java/org/geotools/tpk/TPKZoomLevelV2.java: warnings found and -Werror specified
```

Has an easy fix, use ``Double.valueOf(x)`` instead of the constructor

**Comment by leif81** [comment]
@MurrayMcDonald16 there's one more Travis build error on openjdk11, see below:

https://travis-ci.org/geotools/geotools/builds/603924949?utm_source=github_status&utm_medium=notification

> [INFO] --- maven-compiler-plugin:3.8.0:compile (default-compile) @ gt-tpk ---
> [INFO] Changes detected - recompiling the module!
> [INFO] Compiling 9 source files to /home/travis/build/geotools/geotools/modules/unsupported/tpk/target/classes
> [INFO] -------------------------------------------------------------
> [WARNING] COMPILATION WARNING : 
> [INFO] -------------------------------------------------------------
> [WARNING] /home/travis/build/geotools/geotools/modules/unsupported/tpk/src/main/java/org/geotools/tpk/TPKZoomLevelV1.java:[87,26] Double(double) in java.lang.Double has been deprecated
> [INFO] 1 warning
> [INFO] -------------------------------------------------------------
> [INFO] -------------------------------------------------------------
> [ERROR] COMPILATION ERROR : 
> [INFO] -------------------------------------------------------------
> [ERROR] /home/travis/build/geotools/geotools/modules/unsupported/tpk/src/main/java/org/geotools/tpk/TPKZoomLevelV1.java: warnings found and -Werror specified
> [INFO] 1 error

**Comment by leif81** [comment]
@MurrayMcDonald16 the `maven-compiler-plugin` is passing now, but the `maven-pmd-plugin` has an error. Here's the message from the Travis build log:

`[ERROR] Failed to execute goal org.apache.maven.plugins:maven-pmd-plugin:3.11.0:check (default) on project gt-tpk: You have 4 PMD violations. For more details see: /home/travis/build/geotools/geotools/modules/unsupported/tpk/target/pmd.xml -> [Help 1]`

https://travis-ci.org/geotools/geotools/jobs/608914766?utm_medium=notification&utm_source=github_status

**Comment by aaime** [comment]
Any progress on the CLA? I just got the latest update, but did not see one from the code author.

**Comment by leif81** [comment]
I'm hoping to have an opportunity this week to get it signed.

**Comment by leif81** [comment]
@aaime my employer is stuck on this sentence in para 3 of the corporate contricutor agreement and won't sign yet because of it. I'm not expecting you to be a lawyer, but do you understand the general intent of this statement?

> If any entity institutes
> patent litigation against You or any other entity (including a
> cross-claim or counterclaim in a lawsuit) alleging that your
> Contribution, or the Work to which you have contributed, constitutes
> direct or contributory patent infringement, then any patent licenses
> granted to that entity under this Agreement for that Contribution or
> Work shall terminate as of the date such litigation is filed.

**Comment by aaime** [comment]
OSGeo CLA is modeled after the Apache CLA indeed, so you might find info looking for it. There is also a FAQ here: https://wiki.osgeo.org/wiki/Contributor_Agreement

**Comment by leif81** [comment]
@aaime This is stalled on my side. The person responsible with authority for signing the CLA retired from his job quite suddenly in December. And his replacement hasn't been found yet.

Is there any alternative available to get the branch merged in?

**Comment by aaime** [comment]
Not currently, no, but we may discuss relaxing the CLA requirements in the future (no idea what the outcome will be, mind).

**Comment by jodygarnett** [comment]
If it helps the organizing you are making this agreement with is the open source geospatial foundation, which has a lot of prior art in its portfolio (with projects like grass and geotools being quite old).
Licenses were easier when it was just about use. Open source adds reuse. And then patents and trademarks make it complicated.

As far as I know the Apache License Agreement here is far less strict than the GPL license, although perhaps not as carefree as the LGPL license that geotools uses.

**Comment by leif81** [comment]
Quick update on the subject of signing the CLA. Some good news. I've got someone in my chain of command who is now briefed on this subject and is supportive of the work we are doing here regarding contributing back to GeoTools. The next obvious step is getting him or someone above him to sign the CLA. I don't have an ETA yet though.

**Comment by aaime** [comment]
Good to hear, thanks for the update!

**Comment by leif81** [comment]
Good news! 

I've sent a signed corporate CLA to the info@osgeo.org address just now.

**Comment by aaime** [comment]
Nice. Then the merge conflict should be solved, and then we are good to go I guess.

**Comment by leif81** [comment]
@aaime I'm not seeing the merge conflict. I'm not sure how to proceed. Can you share the error you're seeing?

From my view what I see is a green checkmark that says "This branch has no conflicts with the base branch". It sounds good to me?

**Comment by aaime** [comment]
Sorry it was there because I had the "rebase and merge" merge method chosen, it does not have any if I create an explicit merge commit. 

However I see it's breaking the QA checks (and thus, will break every build on the build servers if it gets merged):

```
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-pmd-plugin:3.11.0:check (default) on project gt-tpk: You have 4 PMD violations. For more details see: /home/travis/build/geotools/geotools/modules/unsupported/tpk/target/pmd.xml -> [Help 1]
```

You can learn more about them, including how to run them locally, here:
https://docs.geotools.org/latest/developer/conventions/code/qa.html

**Comment by leif81** [comment]
@MurrayMcDonald16 can you take a look at those 4 PMD violations identified in the above comment. Once those are fixed this can be merged in.

**Comment by MurrayMcDonald16** [comment]
@leif81 @aaime 

I've pushed up mods to fix the four PMD "errors" as well as an additional code fix to deal with TPK files that contain different image types, there is a build error, not sure who needs to fix this??

[INFO] Downloaded from central: https://repo.maven.apache.org/maven2/slide/slide-webdavlib/2.1/slide-webdavlib-2.1.jar (173 kB at 1.8 MB/s)
[ERROR] [ERROR] Some problems were encountered while processing the POMs:
[FATAL] Non-resolvable parent POM for org.geotools:gt-tpk:23-SNAPSHOT: Could not find artifact org.geotools:unsupported:pom:23-SNAPSHOT and 'parent.relativePath' points at wrong local POM @ line 13, column 13
[WARNING] 'dependencies.dependency.version' for com.github.spotbugs:spotbugs-annotations:jar is either LATEST or RELEASE (both of them are being deprecated) @ line 124, column 20
[WARNING] 'build.plugins.plugin.version' for org.apache.maven.plugins:maven-checkstyle-plugin is missing. @ line 227, column 14
[WARNING] 'build.plugins.plugin.version' for org.apache.maven.plugins:maven-pmd-plugin is missing. @ line 263, column 15
 @ 
[ERROR] The build could not read 1 project -> [Help 1]
[ERROR]   
[ERROR]   The project org.geotools:gt-tpk:23-SNAPSHOT (/home/runner/work/geotools/geotools/modules/unsupported/tpk/pom.xml) has 1 error
[ERROR]     Non-resolvable parent POM for org.geotools:gt-tpk:23-SNAPSHOT: Could not find artifact org.geotools:unsupported:pom:23-SNAPSHOT and 'parent.relativePath' points at wrong local POM @ line 13, column 13 ->

**Comment by leif81** [comment]
@aaime 

@MurrayMcDonald16 has fixed the PMD errors. He's also rebased the branch and force pushed. From our perspective this PR is good to go!

**Comment by aaime** [comment]
There is a conflict, I tried to resolve it but I could not push back to the original branch (I guess the authorization to do so was denied when opening the PR, it's on by default but can be checked off).
So I've created a new PR to replace this one: https://github.com/geotools/geotools/pull/2592

**Comment by aaime** [comment]
Green build on the new PR, merged.

**Comment by aaime** [comment]
Ah, also created a ticket, so that this change will get into the release notes: https://osgeo-org.atlassian.net/browse/GEOT-6581

**Comment by leif81** [comment]
Terrific!

In the original PR is a link to a jira ticket.

https://osgeo-org.atlassian.net/projects/GEOT/issues/GEOT-6393

**Pull request closed without merging**